### PR TITLE
Fix for#2688

### LINKED ISSUE #2688: Uploading attachments fails with apostrophe in filename
### Category
- [ ] Enhancement
- [ x] Bug
- [ ] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: [3.10]

Please specify what version(s) of SharePoint you are targeting: [SPO]

*If you are not using the latest release, please update and see if the issue is resolved before submitting an issue.*
Checked - the code is the same in the latest version so the bug will still appear.

### Expected / Desired Behavior / Question
When uploading an attachment to a List Item, if the attachment has an apostrophe (') in the filename, the upload should still work fine.

### Observed Behavior
When uploading an attachment to a List Item, if the attachment has an apostrophe (') in the filename, the upload fails.

### Steps to Reproduce
Call `attachmentFiles.add()`, as in `sp.web.lists.getByTitle(Constants.ListNames.PvAsPCard).items.getById(itemId).attachmentFiles.add(file.name, file);`

### Other Details
I've seen in the codebase that for certain file operations the filename is encoded with `encodePath(url)`, such as in https://github.com/pnp/pnpjs/blob/5a867ddd19d0360ad6f433220f96241cc2635c8b/packages/sp/files/types.ts#L43 . This is not being done for `attachmentFiles.add` (see https://github.com/pnp/pnpjs/blob/version-3/packages/sp/attachments/types.ts#L33), likely the cause of the problem - I'll investigate further and try submit a fix.

### PR DESCRIPTION
**Pull request by HiltonGiesenow**
#### Category
- [x ] Bug fix?
- [ ] New feature?
- [ ] New sample?
- [ ] Documentation update?

#### Related Issues

Fixed #2688 

#### What's in this Pull Request?

Changes `attachmentFiles.add` to also encode the filename to ensure files with apostrophes in the filename (') can be uploaded.

### CODE REVIEW

**Comment by HiltonGiesenow** [comment]
This is my first PR to this repo, and the change is quite small, but I'm struggling to get the tests to run locally, and the checkin tests are failing, all in `GroupSiteManager`. There are 8 separate tests failing (per the test run above), none look related to an attachment filename. Would it be possible for someone to help me with some better guidance on getting my local testing set up (the guide is -very- hard to follow), or run the tests locally to check if the change is ok, and what's blocking on `GroupSiteManager`?

Txs

**Comment by bcameron1231** [comment]
Can you show us the error messages for GroupSiteManager tests? They work for us (with the exception of 1 of them (canUserCrateGroup), which is fixed in 3.16).

**Comment by HiltonGiesenow** [comment]
Sure - they're in the CI build for this PR - https://github.com/pnp/pnpjs/actions/runs/5036108261/jobs/9032031935?pr=2689

**Comment by bcameron1231** [comment]
Ah. I wouldn't worry too much about these in the GitHub Actions, as these sporadically fail within GitHub. 

I would just make sure they run locally. 

run all SP packages
`npm run test -- --packages sp`

run just groupsitemanager
`npm run test -- --packages sp --single groupsitemanager --cleanup`

Thanks again for the contribution!

**Comment by HiltonGiesenow** [comment]
Hey @bcameron1231 - see my original message, I'm struggling to figure out how to even test locally - I'm not quite clear what the "Minimal Path to Awesome" is here...

**Comment by bcameron1231** [comment]
Ah. First things first, you need to set up a local settings file. We have an example settings file in the root of the project. 

Here is a doc that describes this.
https://pnp.github.io/pnpjs/contributing/settings/

This settings file is used during the test harness to authenticate the application to do tests. We only run tests as app permissions, so there may be a couple tests you run, that do fail.

Once you've set up the settings file in your root of your project. Then you can just run the suite of commands (some of which I listed above to do the testing).

**Comment by HiltonGiesenow** [comment]
when you say "App permissions", I'm guessing you mean an Azure AD app? I'm happy to set that up but not sure what specifically it should look like and I couldn't find that in the doc. Does it need all possible SP settings? Can I also use the same app for the graph tests? If so, what scopes does that need?

I'd be happy to help with fleshing out the docs on this...

**Comment by bcameron1231** [comment]
Yes. You can use the same application for graph/sp testing, as long as you've granted the application the permission it needs to SP and Graph.  Yes, if you want all the tests to pass, you'd need to complete the entire settings config. 

I create a single App Registration called PnPjs. I upload the certificate to that application, and then I basically grant it God mode permissions lol (Obviously you'd want to be doing this on your own dev tenants). 

We would welcome all and any contribution to our docs. We fully believe docs can always be improved, We do try to limit how much of our docs are replications of Microsoft docs though, so linking to external Microsoft resources is okay if you find yourself re-creating it.

 This is what we've already done in the settings configuration documentation. https://pnp.github.io/pnpjs/contributing/settings/#graph-value

**Comment by HiltonGiesenow** [comment]
For sure, it totally wouldn't make sense to duplicate worthwhile content (like how to create an Azure app), just would be good to specify e.g. the Graph permissions needed for the tests, is what I meant for example. I'll try battle this out my side the next couple of weeks, and update the docs accordingly.

**Review by bcameron1231 — APPROVED** [review]
